File: apps/gig-crawler-2/src/commands/SyncGigsCommand.ts
import type { SearchPort } from '../ports/SearchPort.js';
import type { ScraperPort } from '../ports/ScraperPort.js';
import type { LLMPort } from '../ports/LLMPort.js';
import type { GigsPort } from '../ports/GigsPort.js';
import type { Gig } from '../models/gig.js';
import { logger } from '../utils/logger.js';
import { getDateRangeQuery } from '../utils/dateUtils.js';

export interface SyncStats {
  searchResults: number;
  filteredUrls: number;
  scrapedUrls: number;
  gigsExtracted: number;
  gigsCreated: number;
  gigsSkipped: number;
  errors: number;
}

export class SyncGigsCommand {
  constructor(
    private readonly search: SearchPort,
    private readonly scraper: ScraperPort,
    private readonly llm: LLMPort,
    private readonly gigs: GigsPort
  ) {}

  async execute(options: { clearExisting?: boolean } = {}): Promise<SyncStats> {
    logger.info('Starting gig sync (two-pass approach)');

    const stats: SyncStats = {
      searchResults: 0,
      filteredUrls: 0,
      scrapedUrls: 0,
      gigsExtracted: 0,
      gigsCreated: 0,
      gigsSkipped: 0,
      errors: 0,
    };

    try {
      // Clear existing gigs if requested
      if (options.clearExisting) {
        logger.info('=== CLEARING EXISTING GIGS ===');
        const deletedCount = await this.gigs.deleteAllGigs();
        logger.info({ deletedCount }, 'Cleared existing gigs');
      }

      // PASS 1: Discovery
      logger.info('=== PASS 1: Discovery ===');

      // Step 1: Search with Brave
      const dateRange = getDateRangeQuery(30);
      const query = `Rock/indie/Folk/dark συναυλίες Αθήνα ${dateRange}`;

      const searchResults = await this.search.search(query, 20);
      stats.searchResults = searchResults.length;
      logger.info({ count: searchResults.length }, 'Found search results');

      if (searchResults.length === 0) {
        logger.warn('No search results found, aborting sync');
        return stats;
      }

      // Step 2: Filter promising URLs with Gemini
      const promisingUrls = await this.llm.filterPromisingUrls(searchResults);
      stats.filteredUrls = promisingUrls.length;
      logger.info({ count: promisingUrls.length }, 'Filtered to promising URLs');

      if (promisingUrls.length === 0) {
        logger.warn('No promising URLs found, aborting sync');
        return stats;
      }

      // PASS 2: Extraction
      logger.info('=== PASS 2: Extraction ===');

      // Step 3: Scrape URLs
      const scrapedContents = await this.scraper.scrapeMany(promisingUrls);
      const successfulScrapes = scrapedContents.filter((sc) => sc.success);
      stats.scrapedUrls = successfulScrapes.length;
      logger.info({ count: successfulScrapes.length }, 'Successfully scraped URLs');

      // Step 4: Extract gigs from ALL scraped pages in one batch call
      let allGigs: Gig[] = [];
      try {
        allGigs = await this.llm.extractGigsFromMultiplePages(scrapedContents);
      } catch (error) {
        logger.error({ error }, 'Failed to extract gigs from batch');
        stats.errors++;
      }

      stats.gigsExtracted = allGigs.length;
      logger.info({ count: allGigs.length }, 'Extracted total gigs');

      if (allGigs.length === 0) {
        logger.warn('No gigs extracted, sync complete');
        return stats;
      }

      // Step 5: Store gigs in Strapi (with deduplication)
      for (const gig of allGigs) {
        try {
          // Check if gig already exists
          const existingGigId = await this.gigs.findGig(gig.title, gig.date);
          if (existingGigId) {
            logger.info({ title: gig.title, date: gig.date }, 'Skipping duplicate gig');
            stats.gigsSkipped++;
            continue;
          }

          // Get or create venue
          const venueId = await this.gigs.getOrCreateVenue(gig.venueName);

          // Create gig
          const gigId = await this.gigs.createGig(gig, venueId);
          logger.info({ id: gigId, title: gig.title, date: gig.date }, 'Created gig');
          stats.gigsCreated++;
        } catch (error) {
          logger.error({ gig: gig.title, error }, 'Failed to store gig');
          stats.errors++;
        }
      }

      logger.info('=== Sync Complete ===');
      logger.info(
        {
          searchResults: stats.searchResults,
          filteredUrls: stats.filteredUrls,
          scrapedUrls: stats.scrapedUrls,
          gigsExtracted: stats.gigsExtracted,
          gigsCreated: stats.gigsCreated,
          gigsSkipped: stats.gigsSkipped,
          errors: stats.errors,
        },
        'Sync statistics'
      );

      return stats;
    } catch (error) {
      logger.error({ error }, 'Fatal error during sync');
      stats.errors++;
      throw error;
    }
  }
}

File: apps/gig-crawler-2-python-backup/src/commands/sync_gigs_command.py
"""Sync gigs command - orchestrates two-pass workflow."""
from typing import Dict, Any
from ..ports.search_port import SearchPort
from ..ports.scraper_port import ScraperPort
from ..ports.llm_port import LLMPort
from ..ports.gigs_port import GigsPort
from ..utils.logger import setup_logger
from ..utils.date_utils import get_date_range_query

logger = setup_logger(__name__)


class SyncGigsCommand:
    """Command to sync gigs using two-pass approach."""

    def __init__(
        self,
        search: SearchPort,
        scraper: ScraperPort,
        llm: LLMPort,
        gigs: GigsPort,
    ):
        """Initialize command with dependencies.

        Args:
            search: Search adapter (Brave)
            scraper: Scraper adapter (trafilatura)
            llm: LLM adapter (Gemini)
            gigs: Gigs repository (Strapi)
        """
        self.search = search
        self.scraper = scraper
        self.llm = llm
        self.gigs = gigs

    async def execute(self) -> Dict[str, Any]:
        """Execute two-pass sync workflow.

        Returns:
            Dictionary with sync statistics
        """
        logger.info("Starting gig sync (two-pass approach)")

        stats = {
            "search_results": 0,
            "filtered_urls": 0,
            "scraped_urls": 0,
            "gigs_extracted": 0,
            "gigs_created": 0,
            "gigs_skipped": 0,
            "errors": 0,
        }

        try:
            # PASS 1: Search → Filter URLs
            logger.info("=== PASS 1: Discovery ===")

            # Step 1: Search with Brave
            date_range = get_date_range_query(days_ahead=30)
            query = f"live music events concerts gigs Athens Greece {date_range}"

            search_results = await self.search.search(query, count=20)
            stats["search_results"] = len(search_results)
            logger.info(f"Found {len(search_results)} search results")

            if not search_results:
                logger.warning("No search results found, aborting sync")
                return stats

            # Step 2: Filter promising URLs with Gemini
            promising_urls = await self.llm.filter_promising_urls(search_results)
            stats["filtered_urls"] = len(promising_urls)
            logger.info(f"Filtered to {len(promising_urls)} promising URLs")

            if not promising_urls:
                logger.warning("No promising URLs found, aborting sync")
                return stats

            # PASS 2: Scrape → Extract → Store
            logger.info("=== PASS 2: Extraction ===")

            # Step 3: Scrape URLs with trafilatura
            scraped_contents = await self.scraper.scrape_many(promising_urls)
            successful_scrapes = [sc for sc in scraped_contents if sc.success]
            stats["scraped_urls"] = len(successful_scrapes)
            logger.info(f"Successfully scraped {len(successful_scrapes)} URLs")

            # Step 4: Extract gigs from each scraped page
            all_gigs = []
            for scraped_content in successful_scrapes:
                try:
                    gigs = await self.llm.extract_gigs_from_content(scraped_content)
                    all_gigs.extend(gigs)
                except Exception as e:
                    logger.error(f"Failed to extract gigs from {scraped_content.url}: {e}")
                    stats["errors"] += 1

            stats["gigs_extracted"] = len(all_gigs)
            logger.info(f"Extracted {len(all_gigs)} total gigs")

            if not all_gigs:
                logger.warning("No gigs extracted, sync complete")
                return stats

            # Step 5: Store gigs in Strapi (with deduplication)
            for gig in all_gigs:
                try:
                    # Check if gig already exists
                    existing_gig_id = await self.gigs.find_gig(gig.title, gig.date)
                    if existing_gig_id:
                        logger.info(f"Skipping duplicate gig: {gig.title} on {gig.date}")
                        stats["gigs_skipped"] += 1
                        continue

                    # Get or create venue
                    venue_id = await self.gigs.get_or_create_venue(gig.venue_name)

                    # Create gig
                    gig_id = await self.gigs.create_gig(gig, venue_id)
                    logger.info(f"Created gig {gig_id}: {gig.title} on {gig.date}")
                    stats["gigs_created"] += 1

                except Exception as e:
                    logger.error(f"Failed to store gig '{gig.title}': {e}")
                    stats["errors"] += 1

            logger.info("=== Sync Complete ===")
            logger.info(f"Search results: {stats['search_results']}")
            logger.info(f"Filtered URLs: {stats['filtered_urls']}")
            logger.info(f"Scraped URLs: {stats['scraped_urls']}")
            logger.info(f"Gigs extracted: {stats['gigs_extracted']}")
            logger.info(f"Gigs created: {stats['gigs_created']}")
            logger.info(f"Gigs skipped (duplicates): {stats['gigs_skipped']}")
            logger.info(f"Errors: {stats['errors']}")

            return stats

        except Exception as e:
            logger.error(f"Fatal error during sync: {e}")
            stats["errors"] += 1
            raise
